src/lxc/criu.c

```diff
@@ -652,6 +652,106 @@ static void exec_criu(struct criu_opts *opts)
 	free(argv);
 }
 
+/*
+ * Function to check if the checks activated in 'features_to_check' are
+ * available with the current architecture/kernel/criu combination.
+ *
+ * Parameter features_to_check is a bit mask of all features that should be
+ * checked (see feature check defines in lxc/lxccontainer.h).
+ *
+ * If the return value is true, all requested features are supported. If
+ * the return value is false the features_to_check parameter is updated
+ * to reflect which features are available. '0' means no feature but
+ * also that something went totally wrong.
+ *
+ * Some of the code flow of criu_version_ok() is duplicated and maybe it
+ * is a good candidate for refactoring.
+ */
+bool __criu_check_feature(uint64_t *features_to_check)
+{
+	pid_t pid;
+	uint64_t current_bit = 0;
+	int ret;
+	int features = *features_to_check;
+	/* Feature checking is currently always like
+	 * criu check --feature <feature-name>
+	 */
+	char *args[] = { "criu", "check", "--feature", NULL, NULL };
+
+	if ((features & ~FEATURE_MEM_TRACK & ~FEATURE_LAZY_PAGES) != 0) {
+		/* There are feature bits activated we do not understand.
+		 * Refusing to answer at all */
+		*features_to_check = 0;
+		return false;
+	}
+
+	while (current_bit < sizeof(uint64_t) * 8) {
+		/* only test requested features */
+		if (!(features & (1ULL << current_bit))) {
+			/* skip this */
+			current_bit++;
+			continue;
+		}
+
+		pid = fork();
+		if (pid < 0) {
+			SYSERROR("fork() failed");
+			*features_to_check = 0;
+			return false;
+		}
+
+		if (pid == 0) {
+			if ((1ULL << current_bit) == FEATURE_MEM_TRACK)
+				/* This is needed for pre-dump support, which
+				 * enables pre-copy migration. */
+				args[3] = "mem_dirty_track";
+			else if ((1ULL << current_bit) == FEATURE_LAZY_PAGES)
+				/* CRIU has two checks for userfaultfd support.
+				 *
+				 * The simpler check is only for 'uffd'. If the
+				 * kernel supports userfaultfd without noncoop
+				 * then only process can be lazily restored
+				 * which do not fork. With 'uffd-noncoop'
+				 * it is also possible to lazily restore processes
+				 * which do fork. For a container runtime like
+				 * LXC checking only for 'uffd' makes not much sense. */
+				args[3] = "uffd-noncoop";
+			else
+				exit(1);
+
+			null_stdfds();
+
+			execvp("criu", args);
+			SYSERROR("Failed to exec \"criu\"");
+			exit(1);
+		}
+
+		ret = lxc_wait_for_pid_status(pid);
```

Comment:
I think you should use `wait_for_pid(pid)` which will return `0` on success and `-1` on error. The function you're calling right now returns the status that `waitpid(pid, &status, ...)` stores in `status` and this needs to be inspected with appropriate macros.